checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java

```diff
@@ -0,0 +1,199 @@
+    super(checker);
+  }
+
+  @Override
+  public Void visitReturn(ReturnTree node, Void p) {
+    // Only check return types if ownership is being transferred.
+    if (!checker.hasOption(MustCallChecker.NO_LIGHTWEIGHT_OWNERSHIP)) {
+      MethodTree enclosingMethod = TreePathUtil.enclosingMethod(this.getCurrentPath());
+      // enclosingMethod is null if this return site is inside a lambda. TODO: handle lambdas more
+      // precisely?
+      if (enclosingMethod != null) {
+        ExecutableElement methodElt = TreeUtils.elementFromDeclaration(enclosingMethod);
+        AnnotationMirror notOwningAnno = atypeFactory.getDeclAnnotation(methodElt, NotOwning.class);
+        if (notOwningAnno != null) {
+          // Skip return type subtyping check, because not-owning pointer means Object Construction
+          // Checker won't check anyway.
+          return null;
+        }
+      }
+    }
+    return super.visitReturn(node, p);
+  }
+
+  @Override
+  public boolean isValidUse(
+      AnnotatedDeclaredType declarationType, AnnotatedDeclaredType useType, Tree tree) {
+    // MustCallAlias annotations are always permitted on type uses, because these will be validated
+    // by the Object Construction Checker's -AcheckMustCall algorithm.
+    if (!checker.hasOption(MustCallChecker.NO_RESOURCE_ALIASES)) {
+      useType.removeAnnotationByClass(MustCallAlias.class);
+    }
+    return super.isValidUse(declarationType, useType, tree);
+  }
+
+  @Override
+  protected boolean skipReceiverSubtypeCheck(
+      MethodInvocationTree node,
+      AnnotatedTypeMirror methodDefinitionReceiver,
+      AnnotatedTypeMirror methodCallReceiver) {
+    // It does not make sense for receivers to have must-call obligations. If the receiver of a
+    // method were to have a non-empty must-call obligation, then actually this method should
+    // be part of the must-call annotation on the class declaration! So skipping this check is
+    // always sound.
+    return true;
+  }
+
+  /**
+   * Mark (using the extraArgs parameter) any assignments where the LHS is a resource variable, so
+   * that close doesn't need to be considered. See {@link
+   * #commonAssignmentCheck(AnnotatedTypeMirror, AnnotatedTypeMirror, Tree, String, Object...)} for
+   * the code that uses and removes the marks.
+   */
+  @Override
+  protected void commonAssignmentCheck(
+      Tree varTree,
+      ExpressionTree valueExp,
+      @CompilerMessageKey String errorKey,
+      Object... extraArgs) {
+    if (TreeUtils.elementFromTree(varTree).getKind() == ElementKind.RESOURCE_VARIABLE) {
+      // Use the extraArgs array to signal to later stages of the common assignment check that this
+      // is in a resource variable context.
+      Object[] newExtraArgs = Arrays.copyOf(extraArgs, extraArgs.length + 1);
+      newExtraArgs[newExtraArgs.length - 1] = ElementKind.RESOURCE_VARIABLE;
+      super.commonAssignmentCheck(varTree, valueExp, errorKey, newExtraArgs);
```

Comment:
Since the call to commonAssignmentCheck is made right away, how about adding a boolean instance field to this class, setting it to true right before this call, and restoring it to false right after this call?  That is a bit gross, but less so than the change to `extraArgs`.  (As a side benefit, it is more efficient since it does not require array creation and copying.)